### ISSUE TITLE
0.9.8 Electron beam mission does not progress

### ISSUE THREAD

**Issue by endlesssteve** [endless-sky/endless-sky#2961]
FW mission to capture one of the new electron beam weapons from a gunship.
I was able to get an electron beam turret and bring it back. However there was no dialog etc when I landed. Even going to the space port did not trigger anything. At the outfitter the turret is clearly shown in cargo. I saved the game at this point to be able to duplicate or send a save but on restarting the game I got a pop up with the message that technicians had taken the turret out of cargo and I should go to the space port. I went to the space port but there was just the normal description. Only after a quit and restarted the game did the mission finally complete.

**Comment by nylki** [comment]
I think you need to capture the regular electron-beam laser, not the turrent version.

**Comment by Vilhelm16** [comment]
I recall a while back (and I mean roughly more than 6 months ago) this
mission being changed so that you could get either to complete the mission
(at the time, I had experienced the same issue myself). I could be wrong,
though.

**Comment by tehhowch** [comment]
The issue steve is referencing is that the mission did not complete upon landing, despite seeing the item in the outfitter (e.g. it wasn't on a ship that was in some other system), yet did complete upon reloading the save. You are correct that either the Electron Beam or the Electron Turret can be obtained to complete the mission.

**Comment by endlesssteve** [comment]
I have a correction. When you land with the electron turret the mission does not proceed. The save and restart clears this. but Freya is only to be found at the space port until after the leave the planet and land again. The first failure is not in any of my saves but the second I have saved and happens every time.
Is the fact you have delivered an electron weapon only checked on landing?  I don"t know so much about mission mechanics yet. Would this be visible in the mission data file?

**Comment by endlesssteve** [comment]
to complete
		not "FW Southern Recon 1B (Turret): done"
	to fail
		has "FW Southern Recon 1B (Turret): done"
	on complete
		fail "FW Southern Recon 1B (Turret)"
		outfit "Electron Beam" -1
		dialog `You hand off the Electron Beam to Freya's science team. They tell you that she's waiting for you in a cafe in the spaceport.`

The mission seems to get stuck here. It"s not so easy to follow. So far i can"t find the end of the mission at the space port.

**Comment by endlesssteve** [comment]
I've edited a save file so that I'm at the start of the mission. I'll run through and try to locate the error in the flow.

**Comment by endlesssteve** [comment]
I've taken this as far as I can and now I need help.
I can reproduce the problem with a saved game. I also have an almost identical saved game but with a normal electron beam and this works.
I can"t see any failure in the free worlds middle.txt file that contains the mission.

However there is a difference in the save files from the games with turret or with normal beam. I can only assume this is coming from within the game code.

the mission status section of the saves is the same:
fw southern recon 1 done
fw southern recon 1 offered
fw southerm recon 1b active
fw southern recon 1b offered

The other part of the mission fw southern recon 1b (turret) is not mentioned.

The save file from the game where I have a turret in cargo has an additional section:

# What you've done:
mission "FW Southern Recon 1B"
	name "Electron Beam"
	description "Steal an Electron Beam from a Navy gunboat and return it to Trinket for Freya to look at. She thought your best bet for finding lone gunboats might be the Wei or Alnasl systems."
	autosave
	landing
	to offer
		"FW Southern Recon 1: done" != 0
	to complete
		"FW Southern Recon 1B (Turret): done" == 0
	to fail
		"FW Southern Recon 1B (Turret): done" != 0
	destination Trinket
	on complete
		dialog
			"You hand off the Electron Beam to Freya's science team. They tell you that she's waiting for you in a cafe in the spaceport."
		outfit "Electron Beam" -1
		fail "FW Southern Recon 1B (Turret)"
	on offer
		log
			"It looks like the Navy has begun using some more powerful beam weapons and torpedoes. Tasked with plundering one of the electron beam weapons so Freya can get a closer look at it."
		conversation
			label 0
			`Freya thanks you for performing the scouting mission, and you give her a copy of your ship's logs to analyze. As she looks through them, almost immediately something catches her eye. "Looks like the Navy is fielding some new weapons," she says. "Some higher-yield torpedoes, and a beam weapon even more powerful than heavy lasers. Those beam weapons in particular worry me."`
				goto 1
			label 1
			choice
				`	"Yes, that kind of negates the advantage I was hoping we'd have due to our new plasma turrets."`
					goto 2
				`	"Does that mean we're not going to attack them after all?"`
					goto 3
			label 2
			`	"Well," she says, "the plasma turrets weren't the only ace up our sleeve, but still, it's frustrating that the Navy still has such an advantage over us technologically."`
			`	"What should we do?" you ask.`
				goto 3
			label 3
			`	"We're going to have to fight them eventually, but first we need as much information as we can get on their new weapon systems. I'd like you to find one of their gunboats that has the new electron beams installed, steal one of them, and bring it back here. I imagine one of them can't be much bigger than twenty or thirty tons, so you ought to be able to find space for it. Think you can do that?"`
				goto 4
			label 4
			choice
				`	"How can I find a gunboat that's all by itself? They're usually a part of a bigger fleet."`
					goto 5
			label 5
			`	She thinks about that for a moment, then says, "The Navy has shifted most of their capital ships to the defecting systems in the south. That means that some of the systems up north will have lighter defenses. You might try Wei or Alnasl. Good luck, Captain."`
				accept
		event "gunboats in Alnasl" 0

As this is the only difference I can find this must be the problem or at least evidence of the problem.

Now I need help from somebody who understands the significance of this.

**Comment by endlesssteve** [comment]
Writing this down now ( as always) something has jumped off the page and smacked me in the face.
to offer
"FW Southern Recon 1: done" != 0           I assume this means the mission is offered if fw southern recon has NOT been done.   !=0 being false ?    The mission should be done before this is offered.

So I still have the question why this is set like this?

**Comment by jafdy** [comment]
That says that for the mission to be offered the the previous mission must have been done a non zero number of times. !=0 being dose not equal zero.

**Comment by endlesssteve** [comment]
So it"s only the information as per the data file in a different format.

I have more clues: I removed the invisible from the data file and can see that the FW southern recon 1b ( turret) mission is not being offered. If you take off and land repeatedly it eventually is and is visible on the map together with the original electron beam mission. Landing after this starts the flow again though to FW southern battle 1.

**Comment by endlesssteve** [comment]
[Major Tom~turret electron.txt](https://github.com/endless-sky/endless-sky/files/1253706/Major.Tom.turret.electron.txt)

Here is a save file if it helps.

**Comment by marphod** [comment]
At a guess: 
`
        "You hand off the Electron Beam to Freya's science team. They tell you that she's waiting for you in a cafe in the spaceport."
        outfit "Electron Beam" -1
        fail "FW Southern Recon 1B (Turret)"
`

Trying to do ` "Electron Beam" -1 ` is going to behave badly if you don't have one.

**Comment by endlesssteve** [comment]
The "Electron Beam" -1  is after on complete and the mission is active so i don't think it's that.

Is there something in the game that stops certain mission being given at the same time?
If I understand the logic of the mission the normal and turret mission should run in parallel.

**Comment by endlesssteve** [comment]
I now have a modified mission that works. However this doesn't fix the underlying problem.
The game will not offer two mission at exactly the same time with the same "to offer". This can stop forking in the story and breaks this mission.

I"ll go for a pull request for the work-around but I still think the underlying issue should be solved or explained. perhaps there is a good reason why this doesn't work but then people writing missions should know.

**Comment by tehhowch** [comment]
The save file you want to provide is the one that has the mission preceding the offering of the Electron Beam/Turret fetch quest:
[Major.Tom.turret.electron~ReadyToComplete.txt](https://github.com/endless-sky/endless-sky/files/1258646/Major.Tom.turret.electron.ReadyToComplete.txt)

The reason the mission fails to offer is because it needs `landing` - right now it only offers if you check the spaceport.

**Comment by endlesssteve** [comment]
As nobody has stepped forward with a fix I"ve continued to investigate and experiment. The issue has therefor kind of evolved and is now messy and many earlier comments are no longer relevant.

Not having an explanation for the root cause I tried trial and error changes to the turret mission.

I"m sure I tried with adding landing and it didn"t work. I"ll try this specifically again.

The fix I have that works offers the turret mission earlier on Longjump. I have managed the first real pull request but it some how changed 121 files and it needs cleaning up.

**Comment by endlesssteve** [comment]
I"ve ran through from FW southern recon 1 with the following config:

mission "FW Southern Recon 1B (Turret)"
	landing
	source Trinket
	destination Trinket
	to offer
		has "FW Southern Recon 1: done"
	to complete
		not "FW Southern Recon 1B: done"
	to fail
		has "FW Southern Recon 1B: done"
	on complete
		fail "FW Southern Recon 1B"

...and this works. My previous try may have been at a point too late and before i had realized what the flow was.
As your change is simpler we can go with that. I've seen the push request so I'll close the issue.

**Comment by tehhowch** [comment]
You should not close the issue until a fix is merged and a milestone with
`(done)` is assigned.

**Comment by endlesssteve** [comment]
Just trying to be tidy.  sniff.

**Comment by endless-sky** [comment]
Fixed by merging #2977.

**Comment by endlesssteve** [comment]
When the music"s over 
turnout the light
turnout the light
